rust/plugin-lib/src/state_cache.rs

```diff
@@ -0,0 +1,517 @@
+                        ctx.state.chunk.truncate(start - off);
+                        ctx.state.chunk.push_str(text);
+                        if let Some((newline_count, tail)) = nlc_tail {
+                            ctx.state.chunk.push_str(&tail);
+                            let new_nl_count = count_newlines(&text);
+                            let nl_delta = new_nl_count.wrapping_sub(newline_count) as isize;
+                            ctx.apply_delta(start, end, new_len, nl_delta);
+                        } else {
+                            ctx.truncate_cache(start);
+                        }
+                    } else {
+                        ctx.state.chunk.clear();
+                        ctx.state.chunk_offset = 0;
+                        ctx.truncate_cache(start);
+                    }
+                } else {
+                    ctx.state.chunk.clear();
+                    ctx.state.chunk_offset = 0;
+                    ctx.truncate_cache(start);
+                }
+                self.handler.update(ctx);
+                Some(Value::from(0i32))
+            }
+            PluginRequest::DidSave { ref path } => {
+                ctx.state.path = Some(path.to_owned());
+                self.handler.did_save(ctx);
+                None
+            }
+        }
+    }
+
+    fn idle(&mut self, peer: plugin_base::PluginCtx, token: usize) {
+        let ctx = PluginCtx {
+            state: &mut self.state,
+            peer: peer,
+        };
+        self.handler.idle(ctx, token);
+    }
+}
+
+pub fn mainloop<H: Handler>(handler: &mut H) {
+    let mut my_handler = MyHandler {
+        handler: handler,
+        state: MyState::default(),
+    };
+    plugin_base::mainloop(&mut my_handler);
+}
+
+impl<'a, S: Default + Clone> PluginCtx<'a, S> {
+
+    pub fn get_path(&self) -> Option<&PathBuf> {
+        match self.state.path {
+            Some(ref p) => Some(p),
+            None => None,
+        }
+    }
+
+    pub fn add_scopes(&self, scopes: &Vec<Vec<String>>) {
+        self.peer.add_scopes(&self.state.view_id, scopes)
+    }
+
+    pub fn update_spans(&self, start: usize, len: usize, spans: &[ScopeSpan]) {
+        self.peer.update_spans(&self.state.view_id, start, len, self.state.rev, spans)
+    }
+
+    /// Determines whether an incoming request (or notification) is pending. This
+    /// is intended to reduce latency for bulk operations done in the background.
+    pub fn request_is_pending(&self) -> bool {
+        self.peer.request_is_pending()
+    }
+
+    /// Schedule the idle handler to be run when there are no requests pending.
+    pub fn schedule_idle(&mut self, token: usize) {
+        self.peer.schedule_idle(token);
+    }
+
+    /// Find an entry in the cache by line num. On return `Ok(i)` means entry
+    /// at index `i` is an exact match, while `Err(i)` means the entry would be
+    /// inserted at `i`.
+    fn find_line(&self, line_num: usize) -> Result<usize, usize> {
```

Comment:
This is a tricky question. In general, my reasoning is that _in general_ cache manipulation methods might call the peer, and methods on `MyState` don't have access to that. Then, I wanted to keep methods together, rather than segregate them on the basis of whether they need peer access (among other things, it's easy to imagine that changing as the protocol evolves).

This feels worth re-examining, as we've discussed on RPC. I'm tempted to just pass the peer around, but that's not very satisfying when the plugin can do things asynchronously (for example, on behalf of things coming in from a language server).

Reply:
There's a definite friction between the desire to have a very functional API, where everything gets passed around via some context, and the business requirements where we really do need async access to the other side of the channel.

In both the python plugin lib and in xi-core we've eventually settled on just having references to the peer stored in a few places.